File: src/main/java/core/procedures/math/Quotient.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigRatio;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

public class Quotient extends AFn {

  public Quotient() {
    super(new FnArgsBuilder().min(2).max(2).mandatory(new Class[]{Long.class, Long.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "quotient";
  }

  @Override
  public Number apply2(Object arg1, Object arg2) {
    /* Special cases */
    if (Utils.INSTANCE.isOne(arg2)) {
      return Utils.INSTANCE.inexactnessTaint((Number)arg1, (Number) arg2);
    }
    if (Utils.INSTANCE.isZero(arg2)) {
      throw new ArithmeticException("quotient: undefined for 0");
    }
    return apply((Number) arg1, (Number) arg2);
  }

  private Number apply(BigDecimal first, BigDecimal second) {
    int scale = Math.max(first.scale(), second.scale());
    if (scale > 0) {
      return first.divide(second, Utils.INSTANCE.getDEFAULT_CONTEXT()).setScale(0, Utils.INSTANCE.getROUNDING_MODE())
                  .setScale(1, Utils.INSTANCE.getROUNDING_MODE());
    } else {
      return first.divideToIntegralValue(second).setScale(scale, Utils.INSTANCE.getROUNDING_MODE());
    }
  }

  private Number apply(BigInteger first, BigInteger second) {
    return first.divide(second);
  }

  private Number apply(Number first, Number second) {
    if (Utils.INSTANCE.isZero(first)) {
      return Utils.INSTANCE.inexactnessTaint(first, second);
    }
    if ((first instanceof BigDecimal) || (second instanceof BigDecimal)) {
      return apply(Utils.INSTANCE.toBigDecimal(first), Utils.INSTANCE.toBigDecimal(second));
    }
    if ((first instanceof BigInteger) || (second instanceof BigInteger)) {
      return apply(Utils.INSTANCE.toBigInteger(first), Utils.INSTANCE.toBigInteger(second));
    }
    if (((first instanceof Double) || (second instanceof Double) || (first instanceof Float) || (second instanceof Float)) &&
        (Utils.INSTANCE.isInteger(first)) && Utils.INSTANCE.isInteger(second)) {

      return Long.valueOf(first.longValue() / second.longValue()).doubleValue();
    }
    if ((first instanceof Double) || (second instanceof Double) ||
        (first instanceof BigRatio) || (second instanceof BigRatio)) {

      return apply(Utils.INSTANCE.toBigDecimal(first), Utils.INSTANCE.toBigDecimal(second));
    }
    return first.longValue() / second.longValue();
  }
}

File: src/main/java/core/procedures/math/Quotient.kt
package core.procedures.math

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Ratio
import core.utils.Utils
import core.utils.Utils.taint

import java.math.BigDecimal
import java.math.BigInteger

open class Quotient : AFn<Any?, Number?>(name = "quotient", isPure = true, arity = Exactly(2),
                          mandatoryArgsTypes = arrayOf(Long::class.javaObjectType, Long::class.javaObjectType)) {

    override operator fun invoke(arg1: Any?, arg2: Any?) = when {
        Utils.isOne(arg2)  -> arg2!! as Number taint arg1!! as Number
        Utils.isZero(arg2) -> throw ArithmeticException("$name: undefined for 0")
        else               -> invoke(arg1!! as Number, arg2!! as Number)
    }

    private operator fun invoke(f: BigDecimal, s: BigDecimal) = maxOf(f.scale(), s.scale()).let {
        when {
            it > 0 -> f.divide(s, Utils.DEFAULT_CONTEXT).setScale(0, Utils.ROUNDING_MODE).setScale(1, Utils.ROUNDING_MODE)
            else   -> f.divideToIntegralValue(s).setScale(it, Utils.ROUNDING_MODE)
        }
    }

    private operator fun invoke(first: Number, second: Number): Number? {
        val (f, s) = Utils.upcast(first, second)
        return when {
            Utils.isZero(f)                    -> s taint f
            f is BigDecimal && s is BigDecimal -> invoke(f, s)
            f is BigInteger && s is BigInteger -> f.divide(s)
            f is Double     && s is Double && Utils.isInteger(f) && Utils.isInteger(s) -> (f.toLong() / s.toLong()).toDouble()
            f is Float      && s is Float  && Utils.isInteger(f) && Utils.isInteger(s) -> (f.toLong() / s.toLong()).toDouble()
            f is Double     && s is Double     -> invoke(Utils.toBigDecimal(f), Utils.toBigDecimal(s))
            f is Float      && s is Float      -> invoke(Utils.toBigDecimal(f), Utils.toBigDecimal(s))
            f is Ratio      && s is Ratio      -> invoke(Utils.toBigDecimal(f), Utils.toBigDecimal(s))
            else                               -> f.toLong() / s.toLong()
        }
    }
}
